### PR TITLE
SOLR-16654: Improve test

### PR DESCRIPTION
**Pull request by dsmiley**
* remove 2 test config files
* revert "solr.tests.loadSolrXml" option recently added
* use EmbeddedSolrServerTestRule for initialization

https://issues.apache.org/jira/browse/SOLR-16654

### CODE REVIEW

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java [hunk 1]
```diff
@@ -26,13 +26,63 @@
...
 import org.apache.solr.util.stats.MetricUtils;
 import org.junit.BeforeClass;
+import org.junit.ClassRule;
 import org.junit.Test;
 
 /** Test for {@link ThinCache}. */
 public class TestThinCache extends SolrTestCaseJ4 {
 
+  @ClassRule public static EmbeddedSolrServerTestRule solrRule = new EmbeddedSolrServerTestRule();
+  public static final String SOLR_NODE_LEVEL_CACHE_XML =
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/search/TestThinCache.java:40]: String literal (vs a separate file) keeps the test with the config its testing.  Constant field allows it to be used in the "TestSolrXml" test too.
Looking forward to Java 17 where we could use multi-line String syntax!  (maybe worth requiring Java 17 for tests?)

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java [hunk 2]
```diff
@@ -26,13 +26,63 @@
...
+import org.apache.solr.util.TestHarness;
 import org.apache.solr.util.stats.MetricUtils;
 import org.junit.BeforeClass;
+import org.junit.ClassRule;
 import org.junit.Test;
 
 /** Test for {@link ThinCache}. */
 public class TestThinCache extends SolrTestCaseJ4 {
 
+  @ClassRule public static EmbeddedSolrServerTestRule solrRule = new EmbeddedSolrServerTestRule();
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/search/TestThinCache.java:39]: This is fairly new; there will be more rules available in time + randomization but not yet.

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java [hunk 3]
```diff
@@ -26,13 +26,63 @@
...
+    solrRule.startSolr(home);
+
+    Path configSet = createTempDir("configSet");
+    copyMinConf(configSet.toFile());
+    // insert a special filterCache configuration
+    Path solrConfig = configSet.resolve("conf/solrconfig.xml");
+    Files.writeString(
+        solrConfig,
+        Files.readString(solrConfig)
+            .replace(
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/search/TestThinCache.java:69]: This is a bit hacky I admit.  I would like to add a utility method that might look something like
 `updateXml(Path, String xpath, String newXml)` where the Xpath would be like "/config/query/filterCache".  ChatGPT wrote one for me :-)

Comment (dsmiley) [reply solr/core/src/test/org/apache/solr/search/TestThinCache.java:69]: When possible, the Config API can be used as is done in `org.apache.solr.search.json.TestJsonRequestWithEdismaxDefType#test`.  But "parentCacheName" isn't supported / is too exotic.  I hate that the Config API is a custom format and only partially implements what we can specify in XML.  I'm torn between embracing it and pretending it doesn't exist.

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java [hunk 4]
```diff
@@ -26,13 +26,63 @@
...
+                "</config>",
+                "<query>\n"
+                    + "<filterCache\n"
+                    + "      class=\"solr.ThinCache\"\n"
+                    + "      parentCacheName=\"myNodeLevelCacheThin\"\n"
+                    + "      size=\"5\"\n"
+                    + "      initialSize=\"5\"/>\n"
+                    + "</query></config>"));
+
+    solrRule.newCollection().withConfigSet(configSet.toString()).create();
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/search/TestThinCache.java:79]: the rule uses collection terminology even if it's really a core.  The point is consistent terminology.  Sometimes I think we should go the other way (standardize on "core") unless collection specific options are needed.  Shrug.

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java [hunk 5]
```diff
@@ -126,35 +176,4 @@ public void testInitCore() throws Exception {
     // for the other node-level cache, simply check that metrics are accessible
     assertEquals(0, nodeMetricsSnapshot.get("CACHE.nodeLevelCache/myNodeLevelCache.size"));
   }
-
-  @BeforeClass
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/search/TestThinCache.java:130]: Please never put lifecycle methods at the bottom!

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java [hunk 6]
```diff
@@ -26,13 +26,63 @@
...
+                    + "      parentCacheName=\"myNodeLevelCacheThin\"\n"
+                    + "      size=\"5\"\n"
+                    + "      initialSize=\"5\"/>\n"
+                    + "</query></config>"));
+
+    solrRule.newCollection().withConfigSet(configSet.toString()).create();
+
+    // legacy; get rid of this someday!
+    h = new TestHarness(solrRule.getCoreContainer());
+    lrf = h.getRequestFactory("/select", 0, 20);
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/search/TestThinCache.java:83]: I need to refresh the vision and discuss/debate it.  Your test here, like most Solr tests, uses `assertU`, `assertQ`, `req`, `doc`, which are implemented today on XML strings and uses TestHarness.  But the vision is to embrace SolrClient.  Nonetheless we need convenience methods to assert via XPath.  I could see a new base test class emerging that has methods named like this and that which are called with the same idioms yet actually have different return types  (definitely not XML strings).

**Review by dsmiley — COMMENTED** [review]
Again, the SOLR-11872 Ditch TestHarness vision is incomplete and we need convenience utilities not yet in place but we can nonetheless reduce the proliferation of test configuration files as I show here.

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java (hunk 1 above)
Comment (epugh) [reply solr/core/src/test/org/apache/solr/search/TestThinCache.java:40]: I like seeing xml next ot the test too, helps me understand what is going on.  I use the tests as a "reference" to figure things out a lot.

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java (hunk 5 above)
Comment (epugh) [reply solr/core/src/test/org/apache/solr/search/TestThinCache.java:130]: Could we have patterns like that caught in precommit or lint checker?    These are the kinds of things that should be checked automatically for us.   not putting lifecycle methods at the bottom isn't an actual pattern that I knew about for example ;-).  We need more support in keeping formatting clean.

**Review by epugh — APPROVED** [review]
This does look nicer!   It would be great to figure out how to rally folks to make the conversion happen faster..   I hate that we still have SolrTestCase and SolrTestCaseHS for example...     I know some folks use linting tools that kick in when a file changes to fail the build.  So if you touch a test, well then you fail the build until it's migrated to these newer patterns..

**Review by magibney — APPROVED** [review]
Thanks for addressing this, and thanks especially for the extra explanation around what was done, and why.

File: solr/core/src/test/org/apache/solr/search/TestThinCache.java (hunk 2 above)
Comment (stillalex) [reply solr/core/src/test/org/apache/solr/search/TestThinCache.java:39]: @dsmiley shouldn't this have also been `final`?

Comment (dsmiley) [reply solr/core/src/test/org/apache/solr/search/TestThinCache.java:39]: I suppose so.

**Pull request merged**